yt_dlp/extractor/tubetugraz.py

```diff
@@ -0,0 +1,349 @@
+        v = try_get(src, getter, expected_type)
+        if v is not None:
+            return v
+
+
+class TubeTuGrazIE(InfoExtractor):
+    IE_DESC = 'tube.tugraz.at'
+    IE_NAME = 'TubeTuGraz'
+
+    _VALID_URL = r"""(?x)^
+        (?:https?://)?tube\.tugraz\.at/paella/ui/(?:
+            (?P<series>browse\.html\?series=)|
+            (?P<episode>watch.html\?id=)
+        )(?P<id>[0-9a-fA-F]{8}-(?:[0-9a-fA-F]{4}-){3}[0-9a-fA-F]{12})
+    $"""
+    _TEST = {
+        "url": "https://tube.tugraz.at/paella/ui/watch.html?id=f2634392-e40e-4ac7-9ddc-47764aa23d40",
+        "md5": "a23a3d5c9aaca2b84932fdba66e17145",
+        "info_dict": {
+            "id": "f2634392-e40e-4ac7-9ddc-47764aa23d40",
+            "ext": "mp4",
+            "title": "#6 (23.11.2017)",
+            "episode": "#6 (23.11.2017)",
+            "series": "[INB03001UF] Einführung in die strukturierte Programmierung",
+            "creator": "Safran C",
+            "duration": 3295818,
+        }
+    }
+
+    _LOGIN_REQUIRED = False
+    _NETRC_MACHINE = None
+
+    _LOGIN_URL = "https://tube.tugraz.at/Shibboleth.sso/Login?target=/paella/ui/index.html"
+    _LOGIN_SUCCESS_URL = "https://tube.tugraz.at/paella/ui/index.html"
+
+    _API_SERIES = "https://tube.tugraz.at/series/series.json"
+    _API_EPISODE = "https://tube.tugraz.at/search/episode.json"
+
+    def _login(self):
+        username, password = self._get_login_info()
+        if username is None:
+            return
+
+        result = self._download_webpage_handle(
+            self._LOGIN_URL, None,
+            note='downloading login page',
+            errnote='unable to fetch login page',
+            fatal=False)
+        if result is False:
+            return
+        else:
+            _, login_page_handle = result
+
+        result = self._download_webpage_handle(
+            login_page_handle.url, None,
+            note='logging in',
+            errnote='unable to log in',
+            fatal=False,
+            data=urlencode_postdata({
+                b"lang": "de",
+                b"_eventId_proceed": "",
+                b"j_username": username,
+                b"j_password": password
+            }),
+            headers={
+                "referer": login_page_handle.url
+            })
+        if result is False:
+            return
+        else:
+            _, result_page_handle = result
+
+        if result_page_handle.url != self._LOGIN_SUCCESS_URL:
+            self.report_warning("unable to login: incorrect password")
+            return
+
+    def _real_initialize(self):
+        self._login()
+
+    def _real_extract(self, url):
+        match = re.match(self._VALID_URL, url)
+
+        id = match.group('id')
+        if match.group('series') is not None:
+            return self._extract_series(id)
+        elif match.group('episode') is not None:
+            return self._extract_episode(id)
+        else:
+            raise ExtractorError("no video found on page")
+
```

Comment:
better to split this into 2 extractors

Reply:
I don't know exactly. Both extractor paths share a lot of code and splitting the two would probably make the code only marginally simpler.

If this should be split, Series should inherit from the Episode extractor, since it uses a lot of the code from it.
Is there precedence in YT-DLP for episodes and playlists being split into multiple extractors?

Reply:
create a base class with the common functions and inherit both extractors from it. The base class is conventionally called `SomeSiteBaseIE`